src/persist-client/src/internal/state_fetcher.rs

```diff
@@ -0,0 +1,373 @@
+                debug!(
+                    "maybe_init lost the CaS race, using current value: {:?}",
+                    live_diffs.last().map(|x| x.seqno),
+                );
+
+                // Clean up the rollup blob that we were trying to reference.
+                let (_, rollup_key) = initial_state.latest_rollup();
+                self.delete_rollup(shard_id, rollup_key).await;
+
+                return self.fetch_latest_state(shard_id, live_diffs).await;
+            }
+        }
+    }
+
+    // WIP panics if hint contains seqnos <= state.seqno
+    pub async fn update_state(
+        &self,
+        state: &mut State<K, V, T, D>,
+        hint: Vec<VersionedData>,
+    ) -> Result<(), CodecMismatch> {
+        assert!(
+            hint.first().map_or(true, |x| x.seqno > state.seqno),
+            "update_state {} called with invalid hint {:?}",
+            state.seqno,
+            hint.iter().map(|x| x.seqno).collect::<Vec<_>>()
+        );
+
+        // If we have at least one live diff that can advance state, apply it
+        // and return.
+        match hint.first() {
+            Some(x) if x.seqno == state.seqno.next() => {
+                self.metrics.state.update_state_fast_path.inc();
+                self.apply_diffs(state, hint.iter());
+                return Ok(());
+            }
+            _ => {
+                self.metrics.state.update_state_slow_path.inc();
+                // Fall through.
+            }
+        }
+        debug!(
+            "update_state didn't hit update_state fast path {} {:?}",
+            state.seqno,
+            hint.iter().map(|x| x.seqno).collect::<Vec<_>>()
+        );
+
+        // Otherwise, we've gc'd the diffs we'd need to advance state to where
+        // hint is at. Use the hint to kick off fetching a new rollup and use
+        // the result of that.
+        self.metrics.state.update_state_slow_path.inc();
+        *state = self.fetch_latest_state(&state.shard_id, hint).await?;
+        Ok(())
+    }
+
+    pub fn apply_diffs<'a, I: IntoIterator<Item = &'a VersionedData>>(
+        &self,
+        state: &mut State<K, V, T, D>,
+        diffs: I,
+    ) {
+        let mut state_seqno = state.seqno;
+        let diffs = diffs.into_iter().filter_map(move |diff| {
+            if diff.seqno == state_seqno {
+                // No-op.
+                return None;
+            }
+            let diff = self.decode_diff(diff);
+            assert_eq!(diff.seqno_from, state_seqno);
+            state_seqno = diff.seqno_to;
+            Some(diff)
+        });
+        state
+            .apply_diffs(&self.metrics, diffs)
+            .expect("state diff should apply cleanly");
+    }
+
+    // WIP Returns all diffs and a rollup at the version == the first diff
+    pub async fn fetch_live_states(
+        &self,
+        shard_id: &ShardId,
+    ) -> Result<(State<K, V, T, D>, Vec<VersionedData>), CodecMismatch> {
```

Comment:
if there's a tolerable way to handle the lifetimes, I think the API / usage would be much cleaner if we returned something like `StateIterator<K, V, T, D>` that internally owns the state + diffs, and produces &States as an iterator

Reply:
done!

there are tricks to impl Iterator over references, but they rely on multiple non-mut refs (long story). since we have to mutate state here, they can't work. luckly, we don't actually have to impl Iterator, `while let Some` is just as good for the two places we use this